crates/core/src/host/wasmtime/wasm_instance_env.rs

```diff
@@ -396,37 +385,23 @@ impl WasmInstanceEnv {
     /// - `(row, row_len)` doesn't decode from BSATN to a `ProductValue`
     ///   according to the `ProductType` that the table's schema specifies.
     #[tracing::instrument(skip_all)]
-    pub fn insert(caller: FunctionEnvMut<'_, Self>, table_id: u32, row: WasmPtr<u8>, row_len: u32) -> RtResult<u16> {
-        let ctx = caller.data().reducer_context();
-        let txn_type = &ctx.txn_type();
-        let db = &ctx.database();
-        let reducer = &ctx.reducer_name().unwrap_or_default().to_owned();
-        let syscall = &AbiCall::Insert;
-
+    pub fn insert(caller: Caller<'_, Self>, table_id: u32, row: WasmPtr<u8>, row_len: u32) -> RtResult<u32> {
         // TODO: Instead of writing this metric on every insert call,
         // we should aggregate and write at the end of the transaction.
-        let _guard = DB_METRICS
-            .wasm_abi_call_duration_sec
-            .with_label_values(txn_type, db, reducer, syscall)
-            .start_timer();
+        let _guard = caller.data().start_abi_call_timer(AbiCall::Insert);
+
+        Self::cvt(caller, AbiCall::Insert, |caller| {
+            let (mem, env) = Self::mem_env(caller);
 
-        Self::cvt(caller, "insert", AbiCall::Insert, |caller, mem| {
             // Read the row from WASM memory into a buffer.
-            let mut row_buffer = mem.read_bytes(&caller, row, row_len)?;
+            let row_buffer = mem.deref_slice_mut(row, row_len)?;
 
             // Insert the row into the DB. We get back the decoded version.
             // Then re-encode and write that back into WASM memory at `row`.
             // We're doing this because of autoinc.
-            let ctx = caller.data().reducer_context();
-            let new_row = caller.data().instance_env.insert(&ctx, table_id.into(), &row_buffer)?;
-            row_buffer.clear();
-            new_row.encode(&mut row_buffer);
-            assert_eq!(
-                row_buffer.len(),
-                row_len as usize,
-                "autoinc'd row is different encoded size from original row"
-            );
-            mem.set_bytes(&caller, row, row_len, &row_buffer)?;
+            let ctx = env.reducer_context();
+            let new_row = env.instance_env.insert(&ctx, table_id.into(), row_buffer)?;
+            new_row.encode(&mut { row_buffer });
```

Comment:
So that it copies `row_buffer` as a temporary instead of mutating the local. It's like the weirdness that comes with `impl io::Read for &[u8]`, basically. Obviously in this scenario we don't care about accessing the original value again anyway, but I still feel this conveys intent better.